api/include/opentelemetry/trace/trace_state.h

```diff
@@ -97,37 +107,168 @@ class TraceState
     }
   };
 
-  // An empty TraceState.
-  TraceState() noexcept : entries_(new Entry[kMaxKeyValuePairs]), num_entries_(0) {}
+  /**
+   * Returns a newly created TraceState parsed from the header provided.
+   * @param header Encoding of the tracestate header defined by
+   * the W3C Trace Context specification https://www.w3.org/TR/trace-context/
+   *  @return Tracestate A new TraceState instance or DEFAULT
+   */
+  static TraceState FromHeader(nostd::string_view header)
+  {
+    TraceState ts;
+
+    std::size_t begin{0};
+    std::size_t end{0};
+    bool invalid_header = false;
+    while (begin < header.size() && ts.num_entries_ < kMaxKeyValuePairs)
+    {
+      // find list-member
+      end = header.find(kMembersSeparator, begin);
+      if (end == std::string::npos)
+      {
+        // last list member. `end` points to end of it.
+        end = header.size() - 1;
+      }
+      else
+      {
+        // `end` points to end of current list member
+        end--;
+      }
+      auto list_member = TrimString(header, begin, end);  // OWS handling
+      if (list_member.size() == 0)
+      {
+        // empty list member, move to next in list
+        begin = end + 2;  // begin points to start of next member
+        continue;
+      }
+      auto key_end_pos = list_member.find(kKeyValueSeparator);
+      if (key_end_pos == std::string::npos)
+      {
+        // Error: invalid list member, return empty TraceState
+        ts.entries_.reset(nullptr);
+        ts.num_entries_ = 0;
+        break;
+      }
+      auto key   = list_member.substr(0, key_end_pos);
+      auto value = list_member.substr(key_end_pos + 1);
+      if (!IsValidKey(key) || !IsValidValue(value))
+      {
+        // invalid header. return empty TraceState
+        ts.entries_.reset(nullptr);
+        ts.num_entries_ = 0;
+        break;
+      }
+      Entry entry(key, value);
+      (ts.entries_.get())[ts.num_entries_] = entry;
+      ts.num_entries_++;
 
-  // Returns false if no such key, otherwise returns true and populates the value parameter with the
-  // associated value.
-  bool Get(nostd::string_view key, nostd::string_view &value) const noexcept
+      begin = end + 2;
+    }
+    return ts;
+  }
+
+  /**
+   * Creates a w3c tracestate header from TraceState object
+   */
+  std::string ToHeader()
   {
-    for (const auto &entry : Entries())
+    std::string header_s;
+    for (size_t count = 0; count < num_entries_; count++)
     {
+      if (count != 0)
+      {
+        header_s.append(",");
+      }
+
+      auto entry = (entries_.get())[count];
+      header_s.append(std::string(entry.GetKey()));
+      header_s.append(std::string(1, kKeyValueSeparator));
```

Comment:
No need to construct a temporary `std::string` here, just pass `kKeyValueSeparator` to `std::string::append`?

Reply:
thanks fixed.